### PR TITLE
Rep 2604 metadata extension

### PR DESCRIPTION
**Pull request by wdschei**

### CODE REVIEW

File: core/src/test/scala/com/rackspace/com/papi/components/checker/GivenAWadlWithMetaTransforms.scala [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+      // THEN access should be allowed;
+      // ELSE access should be forbidden.
+      if (targetInRoles(target, roles, List("admin"))) {
+        // @TODO: Huh???
+        // @TODO: What should the content be?
+        // @TODO: Do we need the new accessIsAllowedWithContent() method?
+        // @TODO: Should we just keep the new badRequestWithStockContent() method since we know it made it all the way through?
+        //it should behave like accessIsAllowed(validator, methodPut, s"/standard/metadata/${target}foo", roles, description)
+        //it should behave like accessIsAllowedWithContent(validator, methodPut, s"/standard/metadata/${target}foo", roles, putContent, description)
+        it should behave like badRequestWithStockContent(validator, methodPut, s"/standard/metadata/${target}foo", roles, description)
```
Comment (wdschei) [inline core/src/test/scala/com/rackspace/com/papi/components/checker/GivenAWadlWithMetaTransforms.scala]: When using `accessIsAllowed()` it returns  400 with "Bad Content: Expecting the root element to be: meta:meta".

File: core/src/test/scala/com/rackspace/com/papi/components/checker/GivenAWadlWithMetaTransforms.scala [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+      // THEN access should be allowed;
+      // ELSE access should be forbidden.
+      if (targetInRoles("", roles, List("admin", "superAdmin"))) {
+        // @TODO: Huh???
+        // @TODO: What should the content be?
+        // @TODO: Do we need the new accessIsAllowedWithContent() method?
+        // @TODO: Should we just keep the new badRequestWithStockContent() method since we know it made it all the way through?
+        //it should behave like accessIsAllowed(validator, methodPut, s"/custom/metadata/${target}foo", roles, description)
+        //it should behave like accessIsAllowedWithContent(validator, methodPut, s"/custom/metadata/${target}foo", roles, putContent, description)
+        it should behave like badRequestWithStockContent(validator, methodPut, s"/custom/metadata/${target}foo", roles, description)
```
Comment (wdschei) [inline core/src/test/scala/com/rackspace/com/papi/components/checker/GivenAWadlWithMetaTransforms.scala]: When using accessIsAllowed() it returns 400 with "Bad Content: Expecting the root element to be: meta:meta".

File: core/src/test/scala/com/rackspace/com/papi/components/checker/GivenAWadlWithMetaTransforms.scala [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+      // @TODO: Huh??? Why doesn't this work?
+      //if (targetInRoles(target, roles, List("admin", "superAdmin"))) {
+      //  it should behave like accessIsAllowed(validator, methodPut, s"/custom/metadata/${target}foo", roles, description)
+      //} else if (target.equals("service:") && targetInRoles("another_role", roles, List.empty)) {
+      //  it should behave like accessIsAllowed(validator, methodPut, s"/custom/metadata/${target}foo", roles, putContent, description)
+      ////////////////////////////////////////
+      // IF the list of roles contains admin OR superAdmin,
+      // THEN access should be allowed;
+      // ELSE access should be forbidden.
+      if (targetInRoles("", roles, List("admin", "superAdmin"))) {
```
Comment (wdschei) [inline core/src/test/scala/com/rackspace/com/papi/components/checker/GivenAWadlWithMetaTransforms.scala]: Aren't the `roles` supposed to have access to the **PUT**?

File: core/src/test/scala/com/rackspace/com/papi/components/checker/GivenAWadlWithMetaTransforms.scala [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+      // @TODO: Huh??? Why doesn't this work?
+      //if (targetInRoles(target, roles, List("admin", "superAdmin"))) {
+      //  it should behave like accessIsAllowed(validator, methodDelete, s"/custom/metadata/${target}foo", roles, description)
+      //} else if (target.equals("service:") && targetInRoles("another_role", roles, List.empty)) {
+      //  it should behave like accessIsAllowed(validator, methodDelete, s"/custom/metadata/${target}foo", roles, putContent, description)
+      ////////////////////////////////////////
+      // IF the list of roles contains admin OR superAdmin,
+      // THEN access should be allowed;
+      // ELSE access should be forbidden.
+      if (targetInRoles("", roles, List("admin", "superAdmin"))) {
```
Comment (wdschei) [inline core/src/test/scala/com/rackspace/com/papi/components/checker/GivenAWadlWithMetaTransforms.scala]: Aren't the `roles` supposed to have access to the **DELETE**?

File: core/src/test/scala/com/rackspace/com/papi/components/checker/GivenAWadlWithMetaTransforms.scala (hunk 1 above)
Comment (RackerWilliams) [inline core/src/test/scala/com/rackspace/com/papi/components/checker/GivenAWadlWithMetaTransforms.scala]: That's the correct behavior.  Check this out here are some examples of what metadata requests look like. https://github.com/rackerlabs/metadata-guide/tree/master/src/docbkx/api_samples

If you build that project you'll get docs which should give you more context on how metadata is supposed to work.

File: core/src/test/scala/com/rackspace/com/papi/components/checker/GivenAWadlWithMetaTransforms.scala (hunk 3 above)
Comment (RackerWilliams) [inline core/src/test/scala/com/rackspace/com/papi/components/checker/GivenAWadlWithMetaTransforms.scala]: Yea, admins are supposed to have access to PUT and DELETE on metadata and metadata/{key}.  But which roles to use for admins depends on what the metadata extension.  Basically any role that uses a pattern of "*" is considered an admin.

**Comment by RackerWilliams** [comment]
Not sure about your testing strategy.  You should be iterating by different configs.  For example withRemoveDups on and withRemoveDups off....etc etc.

So really I see 3 iterations (each with various config options and using multiple metadata configs)

1. A set of assertions with rax:roles disababled... Metadata requests should all fail with 404 etc.
2. rax:roles enabled.  Get Requests to Meta should all work.  Invalid access should return forbidden.
3. rax:roles mask endabled. Get Requests to Meta should work. Invalid access should return 404/405.

File: core/src/test/scala/com/rackspace/com/papi/components/checker/GivenAWadlWithMetaTransforms.scala (hunk 1 above)
Comment (wdschei) [inline core/src/test/scala/com/rackspace/com/papi/components/checker/GivenAWadlWithMetaTransforms.scala]: I compiled the WADL to a Checker with the `-pr` options put that in the test so I could follow it through the DOT. Things started working much better, so this leads me to believe the `Config` produced by `RaxRolesBehaviors.configWithRolesEnabled()` is not the one I need.

Comment (RackerWilliams) [inline core/src/test/scala/com/rackspace/com/papi/components/checker/GivenAWadlWithMetaTransforms.scala]: Remember you want to test with multiple kinds of configs.  See my notes below.

Comment (wdschei) [inline core/src/test/scala/com/rackspace/com/papi/components/checker/GivenAWadlWithMetaTransforms.scala]: Will do.

File: core/src/test/scala/com/rackspace/com/papi/components/checker/GivenAWadlWithMetaTransforms.scala (hunk 4 above)
Comment (wdschei) [inline core/src/test/scala/com/rackspace/com/papi/components/checker/GivenAWadlWithMetaTransforms.scala]: This turned out to be a WADL ordering issue. ???

**Comment by wdschei** [comment]
The failure breakdown is like this:
```
Standard on TOP 64 custom customer => 16 ea / PUT - Roles  \
                                            |     x        |
                96 custom service  => 24 ea \ DEL - Masked /
               ---
               160

 9128 Passed
 1548 FAIL - |||: [S0] 400 : Illegal character in path
  160 FAIL - Ordering
----- -----
10836 TOTAL

Custom on TOP 128 standard billing => 32 ea / PUT - Roles  \
                                            |     x        |
              128 standard service => 32 ea \ DEL - Masked /
              ---
              256

 9032 Passed
 1548 FAIL - |||: [S0] 400 : Illegal character in path
  256 FAIL - Ordering
----- -----
10836 TOTAL
```

**Comment by RackerWilliams** [comment]
You'll need to URL encode input.

**Comment by wdschei** [comment]
The other issue is if the "standard" meta information is above the "custom" meta information, then the "custom" `PUT`'s and `DELETE`'s don't work for anything other than the admin roles (i.e. pattern="*"). The opposite happens if the "custom" is above the "standard".

**Comment by RackerWilliams** [comment]
That's weird. I'll hit you up on vidyo tomorrow.

**Comment by RackerWilliams** [comment]
You need to url encode in the url...not in headers.

**Comment by wdschei** [comment]
I'm only encode the one URL for the RegEx case:
 * https://github.com/wdschei/api-checker/blob/REP-2604_MetadataExtension/core/src/test/scala/com/rackspace/com/papi/components/checker/GivenAWadlWithMetaTransforms.scala#L79

I don't know where the `\u002C` is coming from in the test name.

**Comment by RackerWilliams** [comment]
The order really shouldn't matter.  One easy way to tell is to create a state machine.  Should be the same machine regardless of ordering.   Can you confirm that?

**Comment by wdschei** [comment]
Other than the ordering of the elements in the `.checker`, they are identical.

**Comment by RackerWilliams** [comment]
Can you send me the .checker and .dot files?  I'll take a look.

**Comment by RackerWilliams** [comment]
Also, what exactly is failing?

**Comment by wdschei** [comment]
Yep. I'll ship them.

**Comment by wdschei** [comment]
What is failing is the `PUT`'s and `DELETE`'s that should be allowed are being not allowed/forbidden for the roles defined in the second `rax:metadata` element.

**Comment by wdschei** [comment]
@RackerWilliams The remaining 224 (56x4) failures are due to the Xerces bug you found. They do pass when the `Config.xsdEngine = "SaxonEE"`, as seen here:
* https://github.com/wdschei/api-checker/blob/REP-2604_MetadataExtension/core/src/test/scala/com/rackspace/com/papi/components/checker/GivenAWadlWithMetaTransforms.scala#L185
* https://github.com/wdschei/api-checker/blob/REP-2604_MetadataExtension/core/src/test/scala/com/rackspace/com/papi/components/checker/GivenAWadlWithMetaTransformsSaxonEE.scala#L185

File: core/src/test/scala/com/rackspace/com/papi/components/checker/RaxRolesBehaviors.scala [hunk 5]
```diff
@@ -22,6 +22,8 @@ import javax.servlet.http.{HttpServletRequest, HttpServletResponse}
 import org.mockito.Mockito._
 import org.scalatest.FlatSpec
 
+import scala.xml.NodeSeq
```
Comment (RackerWilliams) [inline core/src/test/scala/com/rackspace/com/papi/components/checker/RaxRolesBehaviors.scala]: Why?

File: RELEASE.md [hunk 6]
```diff
@@ -1,5 +1,14 @@
 # Releases #
 
+## Release 1.1.3 (2015-09-??) ##
+1. Updated the version string on the Wadl2Checker and Wadl2Dot CLI tools.
+1. Added the new `xsd-grammar-transform` (`g`) option to the Wadl2Checker CLI tool in order to transform the XML after validation.
+1. Added support for the `rax:metadata` extension.
+The `rax:metadata` extension can be configured on a WADL as an attribute at the resource level.
+Due to a bug in Xerces that prevents multiple Schema elements with the same `targetNamespace` from operating correctly,
+the `http://docs.rackspace.com/metadata/api` namespace should be considered reserved.
+It can **NOT** be used elsewhere if this extension is enabled.
```
Comment (RackerWilliams) [inline RELEASE.md]: I think simply stating what the metadata extension is is good enough.  This isn't really a place where we list bugs.

**Comment by RackerWilliams** [comment]
Do you mind squashing these commits down a bit.  Too much stuff in here, that's no longer relevant.

File: core/src/test/scala/com/rackspace/com/papi/components/checker/RaxRolesBehaviors.scala (hunk 5 above)
Comment (wdschei) [inline core/src/test/scala/com/rackspace/com/papi/components/checker/RaxRolesBehaviors.scala]: It was needed for one of the interim changes. I'll pull it back out.

File: RELEASE.md (hunk 6 above)
Comment (wdschei) [inline RELEASE.md]: Pulled.

**Comment by wdschei** [comment]
Comments addressed and commits squashed.

**Comment by wdschei** [comment]
Retest this please.

**Comment by RackerWilliams** [comment]
28,208 tests!  Wow, I think you may have gone a bit overboard there...but hey they all pass and they seem reasonable... so looks good.

**Comment by wdschei** [comment]
It was a little overkill, but the nature of the tests were well suited to iterating over all of the permutations for this given test API since it was small. This huge amount of tests led to finding the Xerces bug and making the root cause easier to identify.

**Pull request merged**